canonical/oathkeeper-configurator
# Add juju config and logic for creating traefik routes

This PR adds:
- oathkeeper-related config options (access rules, headers)
- traefik-related config options (root_url, rule)
- config validation
- logic for generating routing rules with forwardauth middleware for charms related via ingress-per-unit interface

The further work will be to support generating rules for non-charmed workloads.

## Testing Instructions

```
charmcraft pack
juju deploy ./oathkeeper-configurator*.charm
juju config oathkeeper-configurator access_rules=@access_rules.json root_url=http://foo.bar/{{juju_model}}-{{juju_unit}}
```
where `access_rules.json` is a file that contains the rules.

The charm will get blocked unless you deploy any charm that supports ingress-per-unit relation, for example prometheus-k8s:
```
juju deploy prometheus-k8s
juju relate prometheus-k8s oathkeeper-configurator
```

This is because the configurator charm requires some configs which come from that relation to set up the routes (model, unit names, host, port).
When the charm status becomes active, you should see the rendered rule in logs. If you scale up prometheus it should also get updated to include multiple units:
```
Generated unit configs: {'http': {'routers': {'juju-prometheus-k8s-0-test1-router': {'rule': 'Host(`foo.bar`)', 'service': 'juju-prometheus-k8s-0-test1-service', 'entryPoints': ['web'], 'middlewares': ['oathkeeper']}, 'juju-prometheus-k8s-1-test1-router': {'rule': 'Host(`foo.bar`)', 'service': 'juju-prometheus-k8s-1-test1-service', 'entryPoints': ['web'], 'middlewares': ['oathkeeper']}}, 'services': {'juju-prometheus-k8s-0-test1-service': {'loadBalancer': {'servers': [{'url': 'http://foo.bar/test1-prometheus-k8s-0'}]}}, 'juju-prometheus-k8s-1-test1-service': {'loadBalancer': {'servers': [{'url': 'http://foo.bar/test1-prometheus-k8s-1'}]}}}}}
```
